krm-functions/lib/nad/v1/nad.go

```diff
@@ -129,68 +129,74 @@ func (r *NadStruct) GetConfigSpec() string {
 	if r == nil {
 		return ""
 	}
-	s, ok, err := r.K.NestedString(ConfigType...)
-	if err != nil {
-		return ""
-	}
-	if !ok {
-		return ""
-	}
-	return s
+	return r.GetStringValue(ConfigType...)
 }
 
-func (r *NadStruct) GetCNIType() string {
+func (r *NadStruct) GetExistingNadConfig() NadConfig {
 	nadConfigStruct := NadConfig{}
 	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
-		return ""
+		return nadConfigStruct
 	}
-	return nadConfigStruct.Plugins[0].Type
+	return nadConfigStruct
 }
 
-func (r *NadStruct) GetVlan() int {
-	nadConfigStruct := NadConfig{}
-	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
-		return 0
+func (r *NadStruct) GetCNIType() string {
+	for i, plugin := range r.GetExistingNadConfig().Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return r.GetExistingNadConfig().Plugins[i].Type
+		}
 	}
-	return nadConfigStruct.Vlan
+	return ""
+}
+
+func (r *NadStruct) GetVlan() int {
+	return r.GetExistingNadConfig().Vlan
 }
 
 func (r *NadStruct) GetNadMaster() string {
-	nadConfigStruct := NadConfig{}
-	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
-		return ""
+	for i, plugin := range r.GetExistingNadConfig().Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return r.GetExistingNadConfig().Plugins[i].Master
+		}
 	}
-	return nadConfigStruct.Plugins[0].Master
+	return ""
```

Comment:
why do we swallow the error, and return with a zerovalue?

Reply:
get outcome can be empty.. But I refactored it to be a bit more clear